### DISCUSSION TITLE
Non-brain images?

### DISCUSSION THREAD

**Discussion by asy51** [jcreinhold/intensity-normalization#67] [Q&A] [answered]
>All algorithms except Z-score (zscore-normalize) and the Piecewise Linear Histogram Matching (nyul-normalize) are specific to images of the brain.

In what ways are the other algorithms specific to images of the brain? What terrible things will happen if they were used to normalize images of extremities + bone mask, for example?

**Comment by jcreinhold** [comment, marked as answer]
Hi @asy51. FCM and LSQ assume there are three main tissue groups in the mask (CSF, GM, and WM), so those would likely perform poorly on images where there are more tissue types. KDE might be fine, but it's finicky regardless, so you should empirically validate that the method works on your images by examining the set of histograms of the normalized images. I'm not sure what WhiteStripe would do—probably perform poorly—same with RAVEL which uses WhiteStripe.

I'd say go ahead and try whatever method you want and empirically validate the results by examining the histograms of the set of images. You want to see something like the histograms on the right-side of the image on the README.

If you do, I'd be curious to hear about the results. Best of luck!

**Comment by asy51** [comment]
Thank you @jcreinhold for the quick response! I ran each normalization method on MRI images + bone masks, with no customization:

![raw](https://user-images.githubusercontent.com/90255040/192571982-0294d7d2-8cd0-4983-82bd-4436bd127c31.png)
![whitestripe](https://user-images.githubusercontent.com/90255040/192571981-e1811673-00b4-47ba-8e94-801d09364e4b.png)
![nyul](https://user-images.githubusercontent.com/90255040/192571985-edeea442-9b57-4d14-b96b-91fedeb95438.png)
![lsq](https://user-images.githubusercontent.com/90255040/192571986-87e95873-4359-492d-b71b-6eab5bae53d6.png)
![kde](https://user-images.githubusercontent.com/90255040/192571987-32ff76ef-1a58-4c26-a1bc-430e339a459a.png)
![fcm](https://user-images.githubusercontent.com/90255040/192571989-92957946-67d7-44b7-8203-894d2d92a0c4.png)
![zscore](https://user-images.githubusercontent.com/90255040/192571993-282dacde-48ea-444a-8802-a85debdbf26e.png)

```
load image cache
100%|█████████████████████████████████████████████████████████████████████████| 195/195 [00:16<00:00, 11.62it/s]
189
raw
100%|████████████████████████████████████████████████████████████████████████| 189/189 [00:00<00:00, 308.94it/s]
saved /home/yua4/ptoa/figs/raw.png
zscore
100%|█████████████████████████████████████████████████████████████████████████| 189/189 [00:13<00:00, 13.72it/s]
saved /home/yua4/ptoa/figs/zscore.png
whitestripe
100%|█████████████████████████████████████████████████████████████████████████| 189/189 [00:39<00:00,  4.73it/s]
saved /home/yua4/ptoa/figs/whitestripe.png
kde
100%|█████████████████████████████████████████████████████████████████████████| 189/189 [00:26<00:00,  7.01it/s]
saved /home/yua4/ptoa/figs/kde.png
fcm
100%|███████████████████████████████████████████████████████████████████████| 189/189 [2:34:50<00:00, 49.16s/it]
saved /home/yua4/ptoa/figs/fcm.png
lsq
fitting...
100%|███████████████████████████████████████████████████████████████████████| 189/189 [2:45:58<00:00, 52.69s/it]
saved /home/yua4/ptoa/figs/lsq.png
nyul
fitting...
  1%|▍                                                                          | 1/189 [00:00<00:54,  3.43it/s]divide by zero encountered in true_divide
invalid value encountered in multiply
100%|█████████████████████████████████████████████████████████████████████████| 189/189 [00:52<00:00,  3.59it/s]
saved /home/yua4/ptoa/figs/nyul.png
ravel
fitting...
All images must be the same size. Resample/co-register the images.
```
The Raw histogram shows two groups, which makes sense since the dataset is sourced from two studies. FCM and LSQ looks to have the best results, but their runtimes are too long! I'm wondering if it's because they are sample-based methods. If I can't speed up the FCM and LSQ methods, I will stick with zscore, even though their minimums don't line up neatly at 0.

**Reply by jcreinhold** [reply]
This is great! Thanks for posting this!

FCM and LSQ fit fuzzy c-means clustering to the intensities in the image, so it will take longer than calculating summary statistics to z-score normalize the images. It's a tradeoff that you'll have to make a judgement call on.

I believe that z-score normalization makes more sense in the scenario you described, regardless of the qualitative appearance of the histograms, because I'm not quite sure what it means to use the FCM method, as it's created, with bone masks on different anatomical structures—FCM/LSQ really are methods specific to the brain.

**Reply by asy51** [reply]
When using z-score normalization, do you know the best way to set the minimum value of each image to 0? When I use a naive sliding (subtract the min value from every voxel) method, the normalization seems to unravel a little bit.

without sliding
![download](https://user-images.githubusercontent.com/90255040/192633390-5291c627-6370-4f5a-adda-ce7a260e5c97.png)
with sliding
![download](https://user-images.githubusercontent.com/90255040/192633423-48ad9a72-934b-4561-88ae-05908414249b.png)

**Reply by jcreinhold** [reply]
Here's one idea: Let $I\ni I_i$ be the $i^{\text{th}}$ image (a $h_i \times w_i \times d_i$ size set of real numbers). 

Then let $m = \min(\\{\min(I_i): i \in \\{1,\ldots,n\\}\\})$ be the minimum value of all the images in your set of images.

Then add $m$ to voxel in each image to yield $\tilde{I}=\\{I_i + m: i \in \\{1,\ldots,n\\}\\}$—the set of Z-score normalized images shifted such that all values in each $\tilde{I}_i$ are positive while maintaining the normalization.

To normalize a new image $I_{n+1}$ (not in $I$) with this same scheme: simply Z-score normalize $I_{n+1}$, add $m$, and set any negative values in the result to 0. The size of the set of negative values will likely be low if the new image is of the same anatomy, same pulse sequence, doesn't contain pathologies not in $I$, etc, so this cutoff will minimally affect the histogram.

**Reply by asy51** [reply]
Using your suggestion with a few sequences (DESS and TSE, respectively, with n=189)
Z-Score norm without sliding (min is less than 0):
![0c80e9e0-3dfc-4679-9f0b-11ef59a26e9f](https://user-images.githubusercontent.com/90255040/193455750-bb4bbac3-fe07-4763-902a-fc13170ae874.png)
If I'm reading your suggestion correctly, this just shifts the histogram so that the global minimum is 0:
![1f1e3ebd-c9a2-4161-b0d8-643cf98f3613](https://user-images.githubusercontent.com/90255040/193455513-ee67e3ef-576b-49fc-bce7-cc435ead7be0.png)
But the minimum of each image is not 0 with this method. I don't know if this is necessarily a problem for downstream usage, but I suspect so.
I was hoping for something like below which nicely sets the min of each image to 0 without sliding, (LSQ method, n=15):
![ec264f4a-0a2a-4c3d-bb5e-2f914a80a27c](https://user-images.githubusercontent.com/90255040/193456053-8b5e317a-f5d9-44dd-89b9-b09d9e366612.png)
which would be my preferred method, if it didn't take so damn long!

**Reply by jcreinhold** [reply]
The minimum of each image won't necessarily be zero with my suggestion, but they'll all be positive and normalized. Why would this cause an issue?

If you think LSQ normalization is the best for your application, and it makes sense to apply it, them why not just let it run over night and use the set of images then? So what if it takes a while—once you have preprocessed images saved to disk, you're good.

**Reply by asy51** [reply]
>The minimum of each image won't necessarily be zero with my suggestion, but they'll all be positive and normalized. Why would this cause an issue?

You're right, it most likely isn't an issue. Eventually I want the preprocessed input range to match the output range of some function (e.g. tanh with range (-1. +1)), so I thought it might be nice for each input to have the same min value.

>If you think LSQ normalization is the best for your application, and it makes sense to apply it, them why not just let it run over night and use the set of images then? So what if it takes a while—once you have preprocessed images saved to disk, you're good.

The LSQ method might be worth saving to disk, as you say. The runtime is only an issue now, while I'm prototyping it with different configurations, etc.
